Updates to the PEP (#9)
TobiasWrigstad/peps: pep-9999.rst

Hashing
=======

Deep immutability opens up the possibility of any freezable object being
hashable, due to its pointer being a stable, unique integer associated
with the object for its entire post-freeze lifetime. However, there are
several complications which should be kept in mind for any future PEPs

> which build on this work at add this functionality.

Reviewer: > Deep immutability opens up the possibility of any freezable object being
hashable, due to its pointer being a stable, unique integer associated
with the object for its entire post-freeze lifetime.

Isn't the point, that the content of the object will remain unchanged, opposed to the pointer?

cpython already assumes that pointers are stable, that's why the default hash implementation uses them.

> However, there are several complications which should be kept in mind for any future PEPs which build on this work at add this functionality.

This tripped me up at first since it seemed to say that there are complications, without listing them. Maybe:

```
There are several difficulties (listed below) which makes allowing frozen
objects to be hashable non trivial. We believe it's better to leave this
for future work, once immutability is fully implemented.
``` 

<!--
Saying that there are several complications without listing them feels a bit weird. What do you think about rephrasing the paragraph like this:

```
Currently, the hash of an object has to remain constant during its lifetime.
This prevents several objects with mutable state to provide a hash
implementation depending on the internal state. Immutability might allow
more objects to be hashable once they're frozen.

We believe that this PEP is a substancial change on its own. Therefore, we
don't aim to explore this avenue here but leave it as possible future work.
It will also be easier to explore such new possibilities once immutability
is fully implemented and can be used for prototypes.
```

-->

Reviewer 2: Good point, that doesn't quite make sense. How is this:

```suggestion
Deep immutability opens up the possibility of any freezable object being
hashable, due to the fixed state of the object graph making it possible to compute
stable hash values over the graph as is the case with `tuple` and `frozenset` . However,
there are several complications (listed below) which should be kept in mind for any future
PEPs which build on this work at add hashability for frozen objects:
```

to be two lists:

.. code-block:: python
  a = [1, 2, 3, 4]
  b = [1, 2, 3, 4]
  assert(hash(a) == hash(b))

There would be a reasonable expectation that this assertion would be true,
as it is for two identically defined tuples. However, without a careful
implementation of `__hash__` and `__eq__` this would not be the case.
Our opinion is that an approach like that used in `tuplehash` is
recommended in order to avoid this behavior.

Decorators of Immutable Functions
=================================

One natural issue that arises from deeply immutable functions is the
state of various objects which are attached to them, such as decorators.
In particular, the case of `lru_cache` is worth investigating. If the cache
is made immutable, then freezing the function has essentially disabled the
functionality of the decorator. This might be the correct and desirable
functionality, from a thread safety perspective! In practice, we see three
potential approaches:

1. The cache is frozen in its state at the point when freeze is called.
Cache misses will result in an immutability exception.
2. Access to the cache is protected by a lock to ensure thread safety

> 3. There is one version of the cache per interpreter

Reviewer: Do we also mention threads to be compatible with NoGil?

```suggestion
3. There is one version of the cache per interpreter or thread
```

Reviewer 2: Slight change in the wording, but good point.

```suggestion
3. There is one version of the cache per interpreter (*i.e.*, the cache is thread local)
```
